File: architecture-skills/design-patterns/examples/interpreter/interpreter.java
// Interpreter pattern in Java.
//
// A tiny arithmetic grammar: literals, variables, +, -, *. A sealed
// interface with one record per grammar rule keeps the class-per-rule
// structure compact; a sentence is a record tree evaluated against a
// map of variable bindings.

import java.util.Map;

public class interpreter {

    sealed interface Expr permits Literal, Variable, Add, Sub, Mul {
        int interpret(Map<String, Integer> context);
    }

    record Literal(int value) implements Expr {
        public int interpret(Map<String, Integer> context) {
            return value;
        }
    }

    record Variable(String name) implements Expr {
        public int interpret(Map<String, Integer> context) {
            return context.get(name);
        }
    }

    record Add(Expr left, Expr right) implements Expr {
        public int interpret(Map<String, Integer> context) {
            return left.interpret(context) + right.interpret(context);
        }
    }

    record Sub(Expr left, Expr right) implements Expr {
        public int interpret(Map<String, Integer> context) {
            return left.interpret(context) - right.interpret(context);
        }
    }

    record Mul(Expr left, Expr right) implements Expr {
        public int interpret(Map<String, Integer> context) {
            return left.interpret(context) * right.interpret(context);
        }
    }

    public static void main(String[] args) {
        // (x + y) * 2 - z
        Expr expression = new Sub(
                new Mul(new Add(new Variable("x"), new Variable("y")), new Literal(2)),
                new Variable("z"));

        Map<String, Integer> first = Map.of("x", 6, "y", 2, "z", 3);
        Map<String, Integer> second = Map.of("x", 1, "y", 1, "z", 10);

        System.out.println("(x + y) * 2 - z with x=6 y=2 z=3   = " + expression.interpret(first));
        System.out.println("(x + y) * 2 - z with x=1 y=1 z=10  = " + expression.interpret(second));
    }
}

File: architecture-skills/design-patterns/examples/interpreter/interpreter.go
// Interpreter pattern in Go.
//
// A tiny arithmetic grammar: literals, variables, +, -, *. A small Expr
// interface with one struct per grammar rule; a sentence is a struct
// tree evaluated against a map of variable bindings.

package main

import "fmt"

type Context map[string]int

type Expr interface {
	Interpret(ctx Context) int
}

type Literal struct {
	Value int
}

func (l Literal) Interpret(Context) int { return l.Value }

type Variable struct {
	Name string
}

func (v Variable) Interpret(ctx Context) int { return ctx[v.Name] }

type Add struct{ Left, Right Expr }

func (a Add) Interpret(ctx Context) int { return a.Left.Interpret(ctx) + a.Right.Interpret(ctx) }

type Sub struct{ Left, Right Expr }

func (s Sub) Interpret(ctx Context) int { return s.Left.Interpret(ctx) - s.Right.Interpret(ctx) }

type Mul struct{ Left, Right Expr }

func (m Mul) Interpret(ctx Context) int { return m.Left.Interpret(ctx) * m.Right.Interpret(ctx) }

func main() {
	// (x + y) * 2 - z
	expression := Sub{
		Left: Mul{
			Left:  Add{Left: Variable{"x"}, Right: Variable{"y"}},
			Right: Literal{2},
		},
		Right: Variable{"z"},
	}

	first := Context{"x": 6, "y": 2, "z": 3}
	second := Context{"x": 1, "y": 1, "z": 10}

	fmt.Printf("(x + y) * 2 - z with x=6 y=2 z=3   = %d\n", expression.Interpret(first))
	fmt.Printf("(x + y) * 2 - z with x=1 y=1 z=10  = %d\n", expression.Interpret(second))
}
